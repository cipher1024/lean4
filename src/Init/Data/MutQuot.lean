prelude
import Init.Core
import Init.Data.List
import Init.System.IO

import Init.Data.Nat.Basic
import Init.Data.Fin.Basic
import Init.Data.UInt
import Init.Data.Repr
import Init.Data.ToString
import Init.Control.Id
import Init.Util

-- #print prefix Quot

universes u v w

namespace Mutable

structure Quot {α : Type u} (r : α → α → Prop) : Type u :=
mk' :: (toQuot : _root_.Quot r)

def Trunc (α : Type u) : Type u := Quot (fun (_ _ : α) => True)

instance {α : Type u} [Inhabited α] (r : α → α → Prop) : Inhabited (Quot r) :=
⟨ ⟨ _root_.Quot.mk r $ arbitrary _ ⟩ ⟩

namespace Quot
variables {α : Type u} {r : α → α → Prop}

/-
override mk, mk', toQuot
-/

unsafe def unsafeMk'  (r : α → α → Prop) (x : _root_.Quot r) : Mutable.Quot r :=
let inst : Inhabited (Mutable.Quot r) := ⟨ ⟨ x ⟩ ⟩;
match unsafeIO $ IO.Prim.mkRef (unsafeCast x : Nat) with
| Except.ok r => unsafeCast r
| Except.error _ => panic!"cannot create new reference"

unsafe def unsafeMk (r : α → α → Prop) (x : α) : Mutable.Quot r :=
unsafeMk' r (Quot.mk r x)

@[implementedBy Mutable.Quot.unsafeMk]
def mk (r : α → α → Prop) (x : α) : Quot r := ⟨ Quot.mk r x  ⟩

attribute [implementedBy Mutable.Quot.unsafeMk'] mk'

unsafe def unsafeLift {r : α → α → Prop} {β : Sort v} (f : α → β)
  (h : ∀ (a b : α), r a b → f a = f b) (x : Mutable.Quot r) : β :=
match unsafeIO $ IO.Prim.Ref.get (unsafeCast x : IO.Ref α) with
| Except.ok val => f val
| Except.error e =>
let inst : Inhabited β := ⟨x.toQuot.lift _ h⟩;
panic!"failed to read reference"

@[implementedBy unsafeLift]
def lift {β : Sort v} (f : α → β)
  (h : ∀ (a b : α), r a b → f a = f b) : Mutable.Quot r → β
| ⟨x⟩ => _root_.Quot.lift f h x

def ind {β : Quot r → Prop}
  (h : ∀ (a : α), β (mk r a)) : ∀ (q : Quot r), β q
| ⟨x⟩ => _root_.Quot.ind h x

def toQuotAux : Mutable.Quot r → _root_.Quot r :=
lift (_root_.Quot.mk r) (λ a b h => Quot.sound h)

attribute [implementedBy Mutable.Quot.toQuotAux] toQuot

def update {β : Sort v} (f : α → PProd β α)
  (ref : IO.Ref (ULift.{v} α)) : IO (ULift.{u} $ PLift.{v} β) := do
⟨x⟩ ← ref.get;
ref.reset;
let ⟨y,x'⟩ := f x;
ref.set ⟨x'⟩;
pure ⟨⟨y⟩⟩

unsafe def unsafeMutate {β : Sort v} (f : α → PProd β α)
  (h : ∀ (a b : α), r a b → (f a).1 = (f b).1)
  (h' : ∀ (a : α), r a (f a).2) (ref : Quot r) : β :=
match unsafeIO (update f $ unsafeCast ref) with
| Except.ok ⟨⟨x⟩⟩ => x
| Except.error _ =>
let inst : Inhabited β := ⟨ref.toQuot.lift _ h⟩;
panic!"failed IO in mutable quotient"

@[implementedBy unsafeMutate]
def mutate {β : Sort v} (f : α → PProd β α)
  (h : ∀ (a b : α), r a b → (f a).1 = (f b).1)
  (h' : ∀ (a : α), r a (f a).2) : Quot r → β :=
lift _ h

end Quot

end Mutable
