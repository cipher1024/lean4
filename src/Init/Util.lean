/-
Copyright (c) 2019 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
-/
prelude
import Init.Data.String.Basic
import Init.Data.ToString

universes u v
/- debugging helper functions -/
@[neverExtract, extern "lean_dbg_trace"]
def dbgTrace {α : Type u} (s : String) (f : Unit → α) : α :=
f ()

/- Display the given message if `a` is shared, that is, RC(a) > 1 -/
@[neverExtract, extern "lean_dbg_trace_if_shared"]
def dbgTraceIfShared {α : Type u} (s : String) (a : α) : α :=
a

@[extern "lean_dbg_sleep"]
def dbgSleep {α : Type u} (ms : UInt32) (f : Unit → α) : α :=
f ()

@[extern c inline "#4"]
unsafe def unsafeCast {α : Type u} {β : Type v} [inh : @& Inhabited β] (a : α) : β := arbitrary _

@[neverExtract, extern "lean_panic_fn"]
constant panic {α : Sort u} [Inhabited α] (msg : String) : α := arbitrary _

@[noinline] private def mkPanicMessage (modName : String) (line col : Nat) (msg : String) : String :=
"PANIC at " ++ modName ++ ":" ++ toString line ++ ":" ++ toString col ++ ": " ++ msg

@[neverExtract, inline] def panicWithPos {α : Sort u} [Inhabited α] (modName : String) (line col : Nat) (msg : String) : α :=
panic (mkPanicMessage modName line col msg)

-- TODO: should be a macro
@[neverExtract, noinline, nospecialize] def unreachable! {α : Sort u} [Inhabited α] : α :=
panic! "unreachable"
