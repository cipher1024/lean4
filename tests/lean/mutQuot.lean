import Init.Data.MutQuot
import Init.Lean.Util.Trace

universes u v w

namespace Examples

variables {α : Type u}

inductive Perm : List α → List α → Prop
| nil : Perm [] []
| trans {xs ys zs : List α} : Perm xs ys → Perm ys zs → Perm xs zs
| swap {x x'} {xs : List α} : Perm (x :: x' :: xs) (x' :: x :: xs)
| skip {x} {xs xs' : List α} : Perm xs xs' → Perm (x :: xs) (x :: xs')

namespace Perm

theorem refl : ∀ {xs : List α}, Perm xs xs :=
@List.rec _ (fun xs => Perm xs xs) Perm.nil
  (fun x xs IH => Perm.skip IH)

theorem of_Eq : ∀ {xs ys : List α}, xs = ys → Perm xs ys
| _, _, Eq.refl _ => Perm.refl

theorem symm : ∀ {xs ys : List α}, Perm xs ys → Perm ys xs :=
@Perm.rec _ (fun xs ys H => Perm ys xs) Perm.nil
  (fun xs ys zs H₀₁ H₁₂ H₁₀ H₂₁ => Perm.trans H₂₁ H₁₀)
  (fun x x' xs => Perm.swap)
  (fun x xs xs' H IH => Perm.skip IH)

variables {ys : List α} {x y : α}

theorem rotate : ∀ {xs : List α} {x}, Perm (x :: xs) (xs ++ [x]) :=
@List.rec _ (fun (xs : List α) => ∀ x, Perm (x :: xs) (xs ++ [x]))
  (fun x => Perm.refl)
  (fun y ys IH x => Perm.trans Perm.swap $ Perm.trans (Perm.skip $ IH _)
                               (Perm.of_Eq (List.consAppend _ _ _).symm))

-- theorem Perm_append_right {ys ys' : List α} (h : Perm ys ys') :
--   ∀ {xs : List α}, Perm (xs ++ ys) (xs ++ ys') :=
-- @List.rec _ (fun (xs : List α) => Perm (xs ++ ys) (xs ++ ys')) h
--   (fun x xs IH =>
--     show Perm _ _ from
--     @Eq.subst _ (fun zs => Perm zs (x :: xs ++ ys')) _ _ (List.consAppend x xs ys) _)

-- #check @Perm.rec
-- #check @Eq.subst
-- #check @List.rec _

-- theorem flip {x xs} : ∀ {ys : List α}, Perm (x :: xs ++ ys) (xs ++ x :: ys) :=
-- let h := @Eq.subst _ (λ ys => Perm ys (xs ++ [x])) (x :: xs) (x :: xs ++ []) (List.append_nil (x :: xs)).symm;
-- -- _
-- @List.rec _ (λ ys => Perm (x :: xs ++ ys) (xs ++ x :: ys))
--   (show Perm (x :: xs ++ []) (xs ++ [x]) from h Perm.swap )
--   _

end Perm

def MRU (α : Type u) := Mutable.Quot $ @Perm α

namespace MRU

def ofList (xs : List α) : MRU α := Mutable.Quot.mk _ xs

variables [DecidableEq α] [HasToString α]

open Lean

def findAux' (x : α) : List α → List α → PProd (Option α) (List α)
| [], xs => dbgTrace "not found" $ fun _ => ⟨none,xs.reverse⟩
| (y :: ys), zs =>
  if x = y then dbgTrace "found" $ fun _ => ⟨some y, y :: List.reverseAux zs ys⟩
           else dbgTrace "loop"  $ fun _ => findAux' ys (y :: zs)

def findAux (x : α) (xs : List α) : PProd (Option α) (List α) :=
let ⟨r,xs'⟩ := findAux' x xs [];
dbgTrace (toString xs') $ fun _ => ⟨r,xs'⟩

theorem foo (x : α) : ∀ (a b : List α), Perm a b → (findAux x a).fst = (findAux x b).fst := sorry

theorem bar' (x : α) : ∀ (a b : List α), Perm (a ++ b.reverse) ((findAux' x a b).snd) := sorry
-- | [], zs => sorry
-- | y :: ys, zs => sorry
  -- Perm.trans (Perm.of_Eq (List.consAppend y ys _)) $
  -- if h : x = y then sorry else sorry
-- Perm.trans (@Eq.subst _ (Perm _) _ _ (ifNeg h).symm (Perm.skip $ bar _ _)) Perm.refl

theorem bar (x : α) (a : List α) : Perm a ((findAux x a).snd) :=
Perm.trans (Perm.of_Eq (List.appendNil a).symm) $
show Perm (a ++ List.nil) ((findAux x a).snd)
from sorry

def find (ls : MRU α) (x : α) : Option α :=
Mutable.Quot.mutate (findAux x) (foo x) (bar x) ls

#eval
let xs : MRU _ := Mutable.Quot.mk _ [1,2,3,4,5,6,7,8,9,10];
let ys := [10,10,10,10,9,10,10,5,10,10,10,9,10];
(ys.map (fun x => dbgTrace (toString x) $ fun _ => find xs x), dbgTrace "> 9" $ fun _ => find xs 9)

end MRU

end Examples
