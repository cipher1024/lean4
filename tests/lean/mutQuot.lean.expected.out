> 9
loop
loop
loop
loop
loop
loop
loop
loop
found
[9, 1, 2, 3, 4, 5, 6, 7, 8, 10]
10
loop
loop
loop
loop
loop
loop
loop
loop
loop
found
[10, 9, 1, 2, 3, 4, 5, 6, 7, 8]
10
found
[10, 9, 1, 2, 3, 4, 5, 6, 7, 8]
10
found
[10, 9, 1, 2, 3, 4, 5, 6, 7, 8]
10
found
[10, 9, 1, 2, 3, 4, 5, 6, 7, 8]
9
loop
found
[9, 10, 1, 2, 3, 4, 5, 6, 7, 8]
10
loop
found
[10, 9, 1, 2, 3, 4, 5, 6, 7, 8]
10
found
[10, 9, 1, 2, 3, 4, 5, 6, 7, 8]
5
loop
loop
loop
loop
loop
loop
found
[5, 10, 9, 1, 2, 3, 4, 6, 7, 8]
10
loop
found
[10, 5, 9, 1, 2, 3, 4, 6, 7, 8]
10
found
[10, 5, 9, 1, 2, 3, 4, 6, 7, 8]
10
found
[10, 5, 9, 1, 2, 3, 4, 6, 7, 8]
9
loop
loop
found
[9, 10, 5, 1, 2, 3, 4, 6, 7, 8]
10
loop
found
[10, 9, 5, 1, 2, 3, 4, 6, 7, 8]
([(some 10), (some 10), (some 10), (some 10), (some 9), (some 10), (some 10), (some 5), (some 10), (some 10), (some 10), (some 9), (some 10)], (some 9))
